rapids-cmake/cpm/detail/pinning_write_file.cmake

```
@@ -0,0 +1,263 @@
+    # cmake-lint: disable=E1120
+    foreach(index RANGE ${entry_count})
+      string(JSON member MEMBER "${data}" ${index})
+      string(JSON existing_value ERROR_VARIABLE dont_have GET "${pinned_json_entry}" ${member})
+      if(dont_have)
+        string(JSON value GET "${data}" ${member})
+        rapids_cpm_pinning_create_and_set_member(pinned_json_entry ${member} ${value})
+      endif()
+    endforeach()
+  endforeach()
+  set(${json_var} "\"${package_name}\" : ${pinned_json_entry}" PARENT_SCOPE)
+endfunction()
+
+#[=======================================================================[.rst:
+rapids_cpm_pinning_pretty_format_json
+-------------------------------------
+
+
+.. versionadded:: v24.04.00
+
+Formats provided JSON to be easily read by humans
+#]=======================================================================]
+function(rapids_cpm_pinning_pretty_format_json _rapids_json_var)
+  set(pretty_json)
+  #[=[
+  - parse each line
+  - trim all leading spaces
+  - exclusive if line contains "}" or "]" decement indenter
+  - print with given indenter
+  - exclusive if line contains "{" or "[" increment indenter
+  ]=]
+  #
+  set(indent "  ")
+  set(indent_len 2)
+  set(current_indent "")
+  string(REPLACE "\n" ";" input "${${_rapids_json_var}}")
+
+  # Needed due to rules around cmake language list expansion:
+  #
+  # during evaluation of an Unquoted Argument. In such contexts, a string is divided into list
+  # elements by splitting on ; characters not following an unequal number of [ and ] characters
+  #
+  string(REPLACE "[" "``" input "${input}")
+  string(REPLACE "]" "~~" input "${input}")
+  foreach(line ${input})
+    string(STRIP "${line}" line)
+    set(push_indent)
+    set(pop_indent)
+    if(line MATCHES "(\\{|``)")
+      set(push_indent TRUE)
+    endif()
+    if(line MATCHES "(\\}|~~)")
+      set(pop_indent TRUE)
+    endif()
+
+    if(pop_indent AND NOT push_indent)
+      string(SUBSTRING "${current_indent}" ${indent_len} -1 current_indent)
+    endif()
+
+    string(APPEND pretty_json "${current_indent}${line}\n")
+
+    if(push_indent AND NOT pop_indent)
+      string(APPEND current_indent "${indent}")
+    endif()
+  endforeach()
+  string(REPLACE "``" "[" pretty_json "${pretty_json}")
+  string(REPLACE "~~" "]" pretty_json "${pretty_json}")
+  set(${_rapids_json_var} "${pretty_json}" PARENT_SCOPE)
+endfunction()
+
+#[=======================================================================[.rst:
+rapids_cpm_pinning_write_file
+-----------------------------
+
+.. versionadded:: v24.04.00
+
+This function will write out the pinned version info to the provided files when we are in the root
+CMakeLists.txt
+#]=======================================================================]
+function(rapids_cpm_pinning_write_file)
+  if(NOT CMAKE_CURRENT_SOURCE_DIR STREQUAL CMAKE_SOURCE_DIR)
+    return()
+  endif()
+
+  find_package(Git QUIET)
+
+  set(_rapids_json
+      [=[
+{
+"packages" : {
```

Can we remove the spaces that precede colons? This happens quite a few places.
```suggestion
"packages": {
```